### DISCUSSION TITLE
JTA with reactive Panache not working properly?

### DISCUSSION THREAD

**Discussion by paloliska** [quarkusio/quarkus#43800] [Q&A] [answered]
Hi, 

I was looking into io.quarkus.mongodb.panache.common.reactive.Panache.withTransaction(..) method: 
```
Panache.startSession()
                    .invoke(s -> s.startTransaction())
                    .invoke(s -> context.putLocal(SESSION_KEY, s))
                    .chain(s -> work.get())
                    .call(() -> commitTransaction())
                    .onFailure().call(() -> abortTransaction())
                    .eventually(() -> Panache.closeSession());
```
There as you can see we are waiting for `work` to be done, than is called `commitTransaction()`.
However `commitTransaction()` is uni with no subscriber as following: 
```
private static Uni<?> commitTransaction() {
        Context context = vertxContext();
        ClientSession current = context.getLocal(SESSION_KEY);
        return toUni(current.commitTransaction());
    }
```
Does this means that this `commitTransaction` uni is not guaranteed to be run exactly after work? We are subscribed to uni `Panache.startSesstion()`, but not to uni `commitTransaction()`.

Are my thoughts correct?

[a comment by an automated account is not shown]

**Comment by loicmathieu** [comment, marked as answer]
Hi,

No, sorry, but the code is correct not your  thoughts ;) 

`call()` like all other operation on Uni returns a new Uni, all operations are inside the chain of Unis which will be resolved by Quarkus as long as you subscribe to the last Uni in the chain. You usually subscribe to it by returning it to the Quarkus runtime (inside a REST endpoint for example).

The `commitTransaction` method returns a Uni that will be passed to the chain of Unis by the `call()` method.
